### DISCUSSION TITLE
channel manager - channel switch is corrupting the  dataset

### DISCUSSION THREAD

**Discussion by solarEdgeIot** [openthread/openthread#9312] [Q&A]
channel manager - channel switch is corrupting the dataset.
we used the 	otChannelManagerSetAutoChannelSelectionEnabled(otInst, true);
and when we are using it in a noisy environment - this causes a channel swipe and this causes the pending dataset to be corrupted.
witch causes a full disconnect from the thread network.
![Untitled (1)](https://github.com/openthread/openthread/assets/57585949/c221d9e3-2eda-4e60-a422-296eaa866d99)

**Comment by jwhui** [comment]
Can you provide the OpenThread git commit that you are using? From the log output, this appears to be a snapshot of OpenThread from a while ago.

Some things that appear strange:
- `channel` values are quite large (e.g. 67 and 176).
- `channelMask` is zero, when it should not be.
- `key` is changing when it should not be changing.

**Reply by avivfox93** [reply]
Hi,
Yes this OpenThread version is quite old, the last commit is df047ed5cbbf1579cea7d729b1aeac927164aa75.
We also tested with more recent version, 0a79c30a202cf0c96d2ce5e37ee01a0546ed40c5, and still got the same result.

As for our investigation it seems as the pending dataset received and saved to flash is very small (31 bytes) which doesn't contain all the information required. When the delay time expired the active dataset is replaced with the small pending dataset and then after reboot we lost all the network information.

**Reply by avivfox93** [reply]
We created a PR for our suggested fix
https://github.com/openthread/openthread/pull/9318

**Reply by abtink** [reply]
Thanks @avivfox93 

I don't think we can get to a partially specified Pending Dataset.
- `ChannelManager` uses `DatasetUpdater` 
- Here is code in `DatasetUpdater` that constructs the Pending Dataset:
https://github.com/openthread/openthread/blob/650e30f81ccfd798409d459a86fb1ed575b5144a/src/core/meshcop/dataset_updater.cpp#L96-L147
- It reads the current active dataset in `error = Get<ActiveDatasetManager>().Read(dataset); `
- And then updates it using` SetForm(requestedDataset)` (which adds/updates new TLVs into it).
- Before saving it as new `PendingDataset`.  

I guess you may be using an older commit. Do you have https://github.com/openthread/openthread/pull/6227?

**Reply by avivfox93** [reply]
@abtink Thanks for the reply,
Some of the devices in our setup includes https://github.com/openthread/openthread/pull/6227 and some don't.
As for my understanding when a device that doesn't include those commits,
The channel manager will send a change channel request with partially specified pending dataset.
And it will break the network regardless of the Openthread version on the receiving side.
Our suggested fix should help in those cases.
We couldn't find any mechanism that addresses this scenario.

**Reply by solarEdgeIot** [reply]
Hi ,
Thanks @abtink  for the replay.
As @avivfox93 mentions a typical thread network can include multiple thread versions devices.
If one of them includes the fix that can make damage all the rest of the devices

should we add a fix for dilling with this situatoin ?
like this:
https://github.com/openthread/openthread/pull/9318

**Reply by abtink** [reply]
Thanks guys.

I understand that #9318 can address this issue, but the conern is that it is not conformant with the Thread specification. The Thread specification only allows the commissioner to provide a partial pending dataset but in general pending dataset from other nodes should be adopted as is. 

I am afraid adding #9318 (i.e., allowing partial dataset and having device populate any missing TLVs in pending dataset) can have consequences in future (e.g. if a TLV is removed/replaced from dataset definition in future versions of spec, devices using #9318 may re-add it incorrectly). 

Here are options I can suggest:
- Update the older devices to include #6227. This would be the best option, as it would ensure that all devices are fully conformant with the Thread specification. However, I understand that this may not be an easy thing to do.
- If you have a way to disable the channel manager on the older devices. Preferably we want one or two devices within the network to enable "channel manager auto selection".
- Add #9318 as a (possibly private) patch with an optional OT_CONFIG (disabled by default) for use by users that may be impacted by this.

**Reply by avivfox93** [reply]
@abtink Thanks for the explanation.
We decided to add optional OPENTHREAD_CONFIG define 2d5f41648d55690cbc49ad3231c7b0edf18ba912
Also, we saw that in case that the commissioner provides partial dataset, the receiving device will fill the missing Tlv's from its own active dataset https://github.com/openthread/openthread/blob/main/src/core/meshcop/dataset_manager_ftd.cpp#L187
So, we wonder what is the difference between the two cases?
Also, there might be a better way to decide when fill the missing Tlv's, like in the case of missing Master Key in the pending dataset. Then we can make sure that we only using the fix on partial dataset that will certainly wipe our network configuration.

We will be glad for your remarks.
